File: app/src/main/java/org/senatov/model/CompareLineItem.java
/*
 * CompareLineItem — one row in the compare list.
 * Carries text + diff status + tree indent for dir mode.
 * Supports expandable directories with disclosure triangle.
 * Iakov Senatov, 2026
 */
package org.senatov.model;

import lombok.Getter;
import lombok.Setter;


@Getter
public final class CompareLineItem {

    private final int lineNumber;
    private final String text;
    private final DiffStatus status;
    private final int indentLevel;
    private final boolean directory;
    private final String relativePath;
    private final long size;
    private final long lastModifiedMs;
    @Setter
    private boolean expanded;

    public enum DiffStatus {
        IDENTICAL,
        MODIFIED,
        ADDED,
        MISSING,
        HEADER
    }


    public CompareLineItem(int lineNumber, String text, DiffStatus status) {
        this(lineNumber, text, status, 0, false, "", 0, 0);
    }


    public CompareLineItem(int lineNumber, String text, DiffStatus status,
                           int indentLevel, boolean directory, String relativePath) {
        this(lineNumber, text, status, indentLevel, directory, relativePath, 0, 0);
    }


    public CompareLineItem(int lineNumber, String text, DiffStatus status,
                           int indentLevel, boolean directory, String relativePath,
                           long size, long lastModifiedMs) {
        this.lineNumber = lineNumber;
        this.text = text;
        this.status = status;
        this.indentLevel = indentLevel;
        this.directory = directory;
        this.relativePath = relativePath;
        this.size = size;
        this.lastModifiedMs = lastModifiedMs;
        this.expanded = false;
    }


    public String formatted() {
        String marker = switch (status) {
            case IDENTICAL -> "  ";
            case MODIFIED  -> "≠ ";
            case ADDED     -> "+ ";
            case MISSING   -> "- ";
            case HEADER    -> "# ";
        };

        String indent = "  ".repeat(indentLevel);
        String disclosure = directory ? (expanded ? "▼ " : "▶ ") : "  ";
        String icon = directory ? "📁 " : "   ";

        if (lineNumber > 0) {
            return String.format("%s%s%s%s%s", marker, indent, disclosure, icon, text);
        }

        return marker + indent + disclosure + icon + text;
    }
}

File: app/src/main/kotlin/org/senatov/model/CompareLineItem.kt
package org.senatov.model

class CompareLineItem(
    val content: LineContent,
    val status: DiffStatus,
    val treeState: TreeDisplayState? = null,
) {
    val lineNumber: Int get() = content.number
    val text: String get() = content.text
    val indentLevel: Int get() = treeState?.details?.location?.depth ?: 0
    val isDirectory: Boolean get() = treeState?.details?.isDirectory ?: false
    val relativePath: String get() = treeState?.details?.location?.relativePath.orEmpty()
    val size: Long get() = treeState?.details?.metadata?.size ?: 0
    val lastModifiedMs: Long get() = treeState?.details?.metadata?.lastModifiedMs ?: 0
    val isExpanded: Boolean get() = treeState?.isExpanded ?: false

    fun formatted(): String {
        val marker = when (status) {
            DiffStatus.IDENTICAL -> "  "
            DiffStatus.MODIFIED -> "≠ "
            DiffStatus.ADDED -> "+ "
            DiffStatus.MISSING -> "- "
            DiffStatus.HEADER -> "# "
        }
        val indent = "  ".repeat(indentLevel)
        val disclosure = if (isDirectory) (if (isExpanded) "▼ " else "▶ ") else "  "
        val icon = if (isDirectory) "📁 " else "   "
        return "$marker$indent$disclosure$icon$text"
    }
}
